fail fast does not stop command execution once the `BEFORE` command fails

```
[BEFORE] [kiegroup/kogito-runtimes] bash -c "if [ ! -z 'quarkus-3' ] && [ -f .ci/environments/update.sh ]; then .ci/environments/update.sh quarkus-3 rewrite push_changes; fi"

[INFO] NOT OK [Executed in 1235095.288113 ms]

[ERROR] The process '/usr/bin/bash' failed with exit code 1

[COMMANDS] [kiegroup/kogito-runtimes] export MVN_CMD=`bash -c "if [ 'true' = 'true' ]; then printf 'deploy  -DaltDeploymentRepository=local::default::file:///home/jenkins/workspace/KIE/kogito/main/nightly.quarkus-3/drools.integration/maven_deploy_dir '; else printf 'install'; fi"`

[INFO] OK [Executed in 6.868972 ms]
```

build-chain should have stopped execution after before failed

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@kie/build-chain-action",
-  "version": "3.3.0",
+  "version": "3.3.1",
   "description": "Library to execute commands based on github projects dependencies.",
   "main": "dist/index.js",
   "author": "",

--- a/src/service/command/execute-command-service.ts
+++ b/src/service/command/execute-command-service.ts
@@ -40,16 +40,8 @@ export class ExecuteCommandService {
     const current = this.getNodeCommands(node, ExecutionPhase.CURRENT, this._configurationService.getNodeExecutionLevel(node));
     const after = this.getNodeCommands(node, ExecutionPhase.AFTER, this._configurationService.getNodeExecutionLevel(node));
     let skipExecution = this._configurationService.skipExecution(node);
-
-    if (!(await this.executeNodePhase(node, before, skipExecution, {...opts, cwd}, result))) {
-      skipExecution = true;
-    }
-
-    if (!(await this.executeNodePhase(node, current, skipExecution, {...opts, cwd}, result))) {
-      skipExecution = true;
-    }
-
-    // don't have to check if after failed since we are returning immediately anyways
+    skipExecution = !(await this.executeNodePhase(node, before, skipExecution, {...opts, cwd}, result)) || skipExecution;
+    skipExecution = !(await this.executeNodePhase(node, current, skipExecution, {...opts, cwd}, result)) || skipExecution;
     await this.executeNodePhase(node, after, skipExecution, {...opts, cwd}, result);
     return result;
   }

--- a/src/service/command/execute-command-service.ts
+++ b/src/service/command/execute-command-service.ts
@@ -39,10 +39,18 @@ export class ExecuteCommandService {
     const before = this.getNodeCommands(node, ExecutionPhase.BEFORE, this._configurationService.getNodeExecutionLevel(node));
     const current = this.getNodeCommands(node, ExecutionPhase.CURRENT, this._configurationService.getNodeExecutionLevel(node));
     const after = this.getNodeCommands(node, ExecutionPhase.AFTER, this._configurationService.getNodeExecutionLevel(node));
-    const skipExecution = this._configurationService.skipExecution(node);
-    result.push(await this.executeCommands(node, before ?? [], skipExecution, {...opts, cwd}));
-    result.push(await this.executeCommands(node, current ?? [], skipExecution, {...opts, cwd}));
-    result.push(await this.executeCommands(node, after ?? [], skipExecution, {...opts, cwd}));
+    let skipExecution = this._configurationService.skipExecution(node);
+
+    if (!(await this.executeNodePhase(node, before, skipExecution, {...opts, cwd}, result))) {
+      skipExecution = true;
+    }
+
+    if (!(await this.executeNodePhase(node, current, skipExecution, {...opts, cwd}, result))) {
+      skipExecution = true;
+    }
+
+    // don't have to check if after failed since we are returning immediately anyways
+    await this.executeNodePhase(node, after, skipExecution, {...opts, cwd}, result);
     return result;
   }
 
@@ -80,6 +88,29 @@ export class ExecuteCommandService {
     return result;
   }
 
+  /**
+   * Executes given phase and adds the result to the result array. Returns true if the phase execution
+   * succeeded otherwise false.
+   */
+  private async executeNodePhase(
+    node: Node, 
+    commands: string[] | undefined, 
+    skipExecution: boolean, 
+    opts: ExecOptions | undefined, 
+    result: ExecuteNodeResult[]
+  ) {
+    const phaseResult = await this.executeCommands(node, commands ?? [], skipExecution, opts);
+    result.push(phaseResult);
+    if (
+      !this._configurationService.failAtEnd() && 
+      phaseResult.executeCommandResults.find(r => r.result === ExecutionResult.NOT_OK)
+    ) {
+      this.logger.info(`${node.project} failed. Won't execute remaining commands and projects`);
+      return false;
+    }
+    return true;
+  }
+
   private nodeExecutionFailed(result: ExecuteNodeResult[]): boolean {
     return !!result.find(res => res.executeCommandResults.find(r => r.result === ExecutionResult.NOT_OK));
   }
@@ -99,7 +130,6 @@ export class ExecuteCommandService {
       this.logger.endGroup();
 
       if (!this._configurationService.failAtEnd() && this.nodeExecutionFailed(currentNodeResult)) {
-        this.logger.info(`${node.node.project} failed. Won't execute remaining projects`);
         return result;
       }
     }
@@ -168,7 +198,6 @@ export class ExecuteCommandService {
         }
         this.logger.endGroup();
         if (!this._configurationService.failAtEnd() && this.nodeExecutionFailed(currentResults[project])) {
-          this.logger.info(`${project} failed. Won't execute remaining projects`);
           return result;
         }
       }

--- a/test/unitary/service/command/execute-command-service.test.ts
+++ b/test/unitary/service/command/execute-command-service.test.ts
@@ -2,7 +2,7 @@ import "reflect-metadata";
 import { ExecuteCommandService } from "@bc/service/command/execute-command-service";
 import { CommandTreatmentDelegator } from "@bc/service/command/treatment/command-treatment-delegator";
 import { CommandExecutorDelegator } from "@bc/service/command/executor/command-executor-delegator";
-import { ExecutionResult } from "@bc/domain/execute-command-result";
+import { ExecuteCommandResult, ExecutionResult } from "@bc/domain/execute-command-result";
 import { ExecutionPhase } from "@bc/domain/execution-phase";
 import { defaultNodeValue } from "@bc/domain/node";
 import { NodeExecution, NodeExecutionLevel } from "@bc/domain/node-execution";
@@ -312,14 +312,15 @@ describe("executeNodeChain", () => {
     
     // last call to execSpy must be for the last node
     expect(execSpy.mock.calls[3][0]).toStrictEqual(nodeChain[3]);
+    promiseSpy.mockRestore();
   });
 
   test.each([
     ["sequential: fail fast", false, false, 1],
     ["sequential: fail at end", false, true, 2],
-    ["parallel: fail fast", true, true, 2],
+    ["parallel: fail fast", true, false, 1],
     ["parallel: fail at end", true, true, 2],
-  ])("%p", async (_title, isParallel, failAtEnd, numExecCalls) => {
+  ])("node: %p", async (_title, isParallel, failAtEnd, numExecCalls) => {
     jest.spyOn(ConfigurationService.prototype, "isParallelExecutionEnabled").mockReturnValueOnce(isParallel);
     jest.spyOn(ConfigurationService.prototype, "failAtEnd").mockReturnValueOnce(failAtEnd);
     const execSpy = jest.spyOn(executeCommandService, "executeNodeCommands");
@@ -346,5 +347,49 @@ describe("executeNodeChain", () => {
     await executeCommandService.executeNodeChain(nodeChain, undefined);
     expect(execSpy).toHaveBeenCalledTimes(numExecCalls);
 
+  });
+
+  test.each([
+    [
+      "before: sequential: fail fast", 
+      false, false, ExecutionResult.NOT_OK, ExecutionResult.OK, 
+      [ExecutionResult.NOT_OK, ExecutionResult.SKIP, ExecutionResult.SKIP]
+    ],
+    [
+      "before: sequential: fail at end", 
+      false, true, ExecutionResult.NOT_OK, ExecutionResult.NOT_OK,
+      [ExecutionResult.NOT_OK, ExecutionResult.NOT_OK, ExecutionResult.OK]
+    ],
+    [
+      "current: parallel: fail fast", 
+      true, false, ExecutionResult.OK, ExecutionResult.NOT_OK,
+      [ExecutionResult.OK, ExecutionResult.NOT_OK, ExecutionResult.SKIP]
+    ],
+    [
+      "current: parallel: fail at end", 
+      true, true, ExecutionResult.OK, ExecutionResult.NOT_OK,
+      [ExecutionResult.OK, ExecutionResult.NOT_OK, ExecutionResult.OK]
+    ],
+  ])("phase: %p", async (_title, isParallel, failAtEnd, beforeResult, currentResult, expectedResult) => {
+    jest.spyOn(ConfigurationService.prototype, "isParallelExecutionEnabled").mockReturnValueOnce(isParallel);
+    jest.spyOn(ConfigurationService.prototype, "failAtEnd").mockReturnValue(failAtEnd);
+    jest.spyOn(executeCommandService, "getNodeCommands").mockReturnValue(["cmd"]);
+    const execSpy = jest.spyOn(executeCommandService, "executeCommand");
+    execSpy.mockResolvedValueOnce({result: beforeResult} as unknown as ExecuteCommandResult);
+    execSpy.mockResolvedValueOnce({result: currentResult} as unknown as ExecuteCommandResult);
+    execSpy.mockResolvedValueOnce({result: ExecutionResult.OK} as unknown as ExecuteCommandResult);
+    
+    const nodeChain: NodeExecution[] = [
+      {
+        node: {
+          ...defaultNodeValue,
+          project: "project1",
+          depth: 0
+        }
+      }
+    ];
+
+    const result = await executeCommandService.executeNodeChain(nodeChain, undefined);
+    expect(result[0].map(r => r.executeCommandResults[0].result)).toStrictEqual(expectedResult);
   });
 });

